feat: Open Telemetry target

joshwestbrook opened the pull request:
Add support for an Open Telemetry target as an alternative to DataDog.

Wasn't sure of the best way to make it so that consumers of this gem don't have to pull in both the `dogstatsd-ruby` and `opentelemetry-*` gems in order to use one of the targets. Open to any suggestions! 😄

joshwestbrook committed:
feat: Open Telemetry target

driv3r reviewed (changes requested):
It can be simplified quite a bit regarding optional gems, as DD statsd was optional from the start anyway, plus you have to initialize the client manually - so there's no automatic setup for this, so we don't actually reference any 3rd party classes.

driv3r commented:
@joshwestbrook also sorry for such a long reply! Just came back from holidays

joshwestbrook committed:
fix: Remove gemfile group, update README, remove explicit requires in targets

joshwestbrook commented on README.md:
```diff
@@ -61,6 +61,23 @@ Given gem provides built in target for Datadog StatsD client, that uses batch op
+
+Given gem provides built in target for OpenTelemetry Metrics SDK, that uses batch operations to publish metrics.
+
+**NOTE** Be sure to have `opentelemetry-metrics-sdk` gem installed.
```
Sure thing! https://github.com/babbel/puma-plugin-telemetry/pull/30/commits/b7aa0f06b71d6800e8915f6b7132f1f7ab7f949d

joshwestbrook commented on lib/puma/plugin/telemetry/targets/datadog_statsd_target.rb:
```diff
@@ -1,5 +1,11 @@
+  require 'datadog/statsd'
+rescue LoadError
+  # Gracefully handle the case when Datadog::Statsd is not installed
+end
```
That makes it way simpler 😄 https://github.com/babbel/puma-plugin-telemetry/pull/30/commits/b7aa0f06b71d6800e8915f6b7132f1f7ab7f949d

joshwestbrook commented:
> @joshwestbrook also sorry for such a long reply! Just came back from holidays

No worries at all! Thanks for taking a look

joshwestbrook committed:
fix: Don't flush metrics when they're recorded by default

stevenharman commented:
As this repo seems… abandoned? I've forked and started adding things we need in production, including Puma 7 support. I'm currently [adding OTel support too](https://github.com/stevenharman/puma-plugin-telemetry_too/pull/3).

amrrbakry committed:
Merge remote-tracking branch 'origin/main' into open-telemetry

# Conflicts:
#	Gemfile.lock
#	README.md

amrrbakry committed:
fix: address review feedback and update for current otel sdk

- Make the open telemetry integration spec parsing nil-safe and
  remove an unused variable.
- Pass force_flush: true in the integration fixture. The metrics-sdk
  gem resolved to 0.15, where nothing pulls the console exporter
  unless the target forces a flush.
- Fix the README options table, document the required meter_provider
  and the force_flush option, and restore the note about installing
  the dogstatsd-ruby gem.
- Reword the changelog entry. The gemspec never had a dogstatsd-ruby
  runtime dependency.

amrrbakry commented on spec/puma/plugin/telemetry/config_spec.rb:
```diff
@@ -45,6 +45,21 @@ module Telemetry
           end
 
+          context 'when built in: Open Telemetry' do
+            let(:meter_provider) { double('otel meter provider', meter: double('otel meter')) }
+
```
This double does not raise: RSpec doubles accept any arguments for stubbed messages unless constrained with `.with`, so `meter_provider.meter('puma.telemetry')` works fine against `double('otel meter provider', meter: ...)`. Verified locally — the spec passes as-is, so we are leaving it unchanged.

amrrbakry committed:
fix: pin opentelemetry gems for ruby 3.0 support

CI tests ruby 3.0, and the newest OpenTelemetry gems require
ruby 3.3. Pin the newest versions that install on ruby 3.0, and
drop the unused otlp exporter gems, which pull in google-protobuf
with its own ruby floor. The fixture only uses the console
exporter.

amrrbakry committed:
feat: default the open telemetry metric prefix to puma

Metric names now get the puma prefix by default, for example
puma.workers.booted. Pass prefix: nil for unprefixed names.

amrrbakry commented:
Thank you so much! Released in https://github.com/babbel/puma-plugin-telemetry/releases/tag/v1.2.0
